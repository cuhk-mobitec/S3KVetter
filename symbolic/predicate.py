# Copyright - see copyright.txt
import codecs
import collections

class Predicate:
	"""Predicate is one specific ``if'' encountered during the program execution.
	   """
	def __init__(self, st, result):
		# symtype means the symbolic type (object). It stores an symbolic object. To get the path condition, you can print symtype.toString()
		self.symtype = st
		# boolean value, i.e., whether the path is true or false
		self.result = result

	def getVars(self):
		return self.symtype.getVars()

	def getVarStr(self):
		return self.symtype.toString()

	def __eq__(self, other):
		if isinstance(other, Predicate):
			if self.result != other.result:
				return False
			else:
				#Boolean variable is None
				if other.symtype.expr != None and other.symtype.expr[0].find('dict_in') == 0:
					return self.symtype.expr[2] == other.symtype.expr[2]
				else:
					return self.symtype.symbolicEq(other.symtype)
		else:
			return False

	def __hash__(self):
		return hash(self.symtype)

	def __str__(self):
		return self.symtype.toString() + " (%s)" % (self.result)

	def __repr__(self):
		return self.__str__()

	def negate(self):
		"""Negates the current predicate"""
		assert(self.result is not None)
		self.result = not self.result

	def isbom(self, expr):
		if isinstance(expr, list):
			for subexpr in expr:
				return self.isbom(subexpr)
		elif isinstance(expr, str):
			if expr.startswith(codecs.BOM_UTF8.decode('utf-8')):
				return True