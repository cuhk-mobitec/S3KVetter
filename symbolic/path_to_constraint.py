# Copyright: see copyright.txt

import logging
import pdb

from .predicate import Predicate
from .constraint import Constraint
from .verify import VerifyEngine
import hashlib



log = logging.getLogger("se.pathconstraint")

class PathToConstraint:
	def __init__(self, add):
		self.constraints = {}
		self.add = add
		self.root_constraint = Constraint(None, None)
		self.current_constraint = self.root_constraint
		self.expected_path = None
		
	def reset(self,expected):
		self.current_constraint = self.root_constraint
		if expected==None:
			self.expected_path = None
		else:
			self.expected_path = []
			tmp = expected
			while tmp.predicate is not None:
				self.expected_path.append(tmp.predicate)
				tmp = tmp.parent

	#Build up constraint
	def whichBranch(self, branch, symbolic_type):
		""" This function acts as instrumentation.
		Branch can be either True or False."""
		# self in fact is SymbolicObject.SI, a class variable.
		
		# add both possible predicate outcomes to constraint (tree)
		# the second time: (== out2#hello, hello) (True)

		#if isinstance(symbolic_type.expr, list):
		#For Boolean operator, symbolic_type.expr is None
		if symbolic_type.expr != None:
			for expr in symbolic_type.expr:
				if isinstance(expr, list):
					op = expr[0]
					if op.find('dict') == 0 and len(symbolic_type.expr) ==3:
						symbolic_type.expr = symbolic_type.expr[1]
						branch = symbolic_type.val
						break
					elif op.find('dict') == 0 and len(symbolic_type.expr) !=3:
						print("do not consider this case yet")
						raise ValueError
				else:
					continue
		# remove the outer layer since it is just a dumpy comparison so that we can enter this function 
		p = Predicate(symbolic_type, branch)
		p.negate()
		cneg = self.current_constraint.findChild(p) # check whether the other branch has been reached or not
		p.negate()
		c = self.current_constraint.findChild(p)

		if c is None:
			c = self.current_constraint.addChild(p)
			# we add the new constraint to the queue of the engine for later processing
			log.debug("New constraint: %s" % c)

			self.add(c)
			
		# check for path mismatch
		# IMPORTANT: note that we don't actually check the predicate is the
		# same one, just that the direction taken is the same
		if self.expected_path != None and self.expected_path != []:
			expected = self.expected_path.pop()
			# while not at the end of the path, we expect the same predicate result
			# at the end of the path, we expect a different predicate result
			done = self.expected_path == []
			if ( not done and expected.result != c.predicate.result or \
				done and expected.result == c.predicate.result ):
				print("Replay mismatch (done=",done,")")
				print(expected)
				print(c.predicate)

		if cneg is not None:
			# We've already processed both branches
			cneg.processed = True
			c.processed = True
			log.debug("Processed constraint: %s" % c)

		self.current_constraint = c

	def toDot(self):
		# print the thing into DOT format
		header = "digraph {\n"
		footer = "\n}\n"
		return header + self._toDot(self.root_constraint) + footer

	def _toDot(self,c):
		if (c.parent == None):
			label = "root"
		else:
			label = c.predicate.symtype.toString()
			if not c.predicate.result:
				label = "Not("+label+")"
		node = "C" + str(c.id) + " [ label=\"" + label + "\" ];\n"
		edges = [ "C" + str(c.id) + " -> " + "C" + str(child.id) + ";\n" for child in c.children if not child.hidden ]

		# The final exection path
		if len(c.children) == 0:
			try:
				label = "returnResult("+str(c.returnResult)+")"
				# use hash to generate a unique node id to act the same purpose of c.id
				# On the other hand, it will merge two nodes to one, if they share the same result.
				node_id = int(hashlib.sha256(str(c.returnResult).encode('utf-8')).hexdigest(), 16) % 10**8
				node = node + "C" + str(node_id) + " [ label=\"" + label + "\" ];\n"
				edges = [ "C" + str(c.id) + " -> " + "C" + str(node_id) + ";\n" ]
			except AttributeError as msg:
				print("handing output returnResult error: ", msg)

		return node + "".join(edges) + "".join([ self._toDot(child) for child in c.children if not child.hidden ])

	@staticmethod
	def balanceBracket(s):
		stack=[]
		bracket,close="(",")" 
		for i in s:
			if i in bracket:
				stack.append(i)
			if i in close:
				if len(stack)==0:
					raise TypeError("the bracket is not balanced", s)
				else:   
					stack.pop()
		if len(stack):
			return len(stack) * ")"
		else:
			return None

	def toSMTLib2(self, solver):
		# convert the path to SMT-lib2 format language
		smtLibExprs = self._toSMTLib2(self.root_constraint, solver).split('\n')
		
		BalancedSmtLibExprs = ""
		for smtLibExpr in smtLibExprs:
			addBrackets = self.balanceBracket(smtLibExpr)
			balanceBrackets = addBrackets  +"\n" if addBrackets != None else  "\n"
			smtLibExpr += balanceBrackets
			BalancedSmtLibExprs += smtLibExpr.lstrip(' ')
		return BalancedSmtLibExprs

	def _toSMTLib2(self, c, solver, pre = ""):
		smtLibExpr = pre
		if c.predicate != None:
			c.predicate.negate()
			predicateExpr = solver.findCVCExpr([], c.predicate)
			# we need to negate the result again. Otherwise, the path True and False would be reversed.
			c.predicate.negate()
			smtLibExpr +=  " (and "+ predicateExpr

		if len(c.children) == 0:
			try:
				if isinstance(c.returnResult, int):
					smtLibExpr += " (= returnVal " + str(c.returnResult) +")\n" 
				elif isinstance(c.returnResult, str):
					smtLibExpr += " (= returnVal " + "\"" + str(c.returnResult) + "\"" +")\n"
				elif isinstance(c.returnResult, bytes):
					print("do not fully support bytes type yet")
					smtLibExpr += " (= returnVal " + str(c.returnResult) +")\n" 
				elif isinstance(c.returnResult, bool):
					smtLibExpr += " (= returnVal " + str(c.returnResult) +")\n"
				elif isinstance(c.returnResult, dict):
					flatDict = dict()
					VerifyEngine.flatNestDict(c.returnResult, flatDict)
					for key,value in flatDict.items():
						smtLibExpr += " (= returnVal." + str(key) +" " 
						if isinstance(value, str):
							smtLibExpr += "\"" +value + "\""  +")"
						elif isinstance(value, int) or isinstance(value, bytes) or isinstance(value, bool):
							smtLibExpr += str(value) +")"
						else:
							smtLibExpr += "\"" + str(value) +"\")"
					smtLibExpr += "\n"
				else:
					smtLibExpr += " (= returnVal " + "\"" + str(c.returnResult) + "\"" +")\n" 
			except AttributeError as msg:
				print("handing output returnResult erroro: ", msg)

		return smtLibExpr + "".join([self._toSMTLib2(child, solver, smtLibExpr) for child in c.children if not child.hidden])


