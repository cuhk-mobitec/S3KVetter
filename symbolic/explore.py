# Copyright: see copyright.txt

from collections import deque
import logging
import os, sys, pdb, traceback
import codecs, utils, functools, copy, re, json
from termcolor import colored

from .z3_wrap import Z3Wrapper
from .path_to_constraint import PathToConstraint
from .invocation import FunctionInvocation
from .symbolic_types import symbolic_type, SymbolicType, getSymbolic, SymbolicDict
from .predicate import Predicate
from .constraint import Constraint

from bdb import BdbQuit


log = logging.getLogger("se.conc")

class ExplorationEngine:
	def __init__(self, funcinv, filename, configPath, solver="z3"):
		self.invocation = funcinv
		# the input to the function
		self.symbolic_inputs = {}  # string -> SymbolicType
		# initialize	
		for n in funcinv.getNames():
			self.symbolic_inputs[n] = funcinv.createArgumentValue(n)

		self._fileName = filename

		self.constraints_to_solve = deque([])
		self.num_processed_constraints = 0

		self.path = PathToConstraint(lambda c : self.addConstraint(c))
		# link up SymbolicObject to PathToConstraint in order to intercept control-flow
		symbolic_type.SymbolicObject.SI = self.path
		if symbolic_type.SymbolicObject.SI == None:
			print("symbolic_type.SymbolicObject.SI is none")

		if solver == "z3":
			self.solver = Z3Wrapper()
		elif solver == "cvc":
			from .cvc_wrap import CVCWrapper
			self.solver = CVCWrapper()
		else:
			raise Exception("Unknown solver %s" % solver)

		# outputs
		self.generated_inputs = []
		self.execution_return_values = []

		self.numofExecution = 0
		self.configPath = configPath
		self.exploreDone = False
		if configPath and os.path.isfile(configPath):
			self.hookFuncs(configPath)

	def hookFuncs(self, configPath):
		with open(configPath) as f:
			configJson = json.load(f)
			for item in configJson['hookedFunc']:
				module = item['module']
				className = item['class']
				api = item['api']
				location = item['location']
				symbolicvar = item['symbolicvar']
				replacedvar = item['replacedvar']
				if (len(symbolicvar) != len(replacedvar)) or (location[0] != 'args' and location[0] != 'kwargs' and location[0] != 'returnVal'):
					print('Incorrect configuration parameters!')
				if item['class'] == '':
					code  = "import %s\n" % module
				else:
					code  = "from %s import %s\n" % (module, className)

				code += "%s = self.processhookFuncs(%s, %s, %s, %s)" % (api, api, location, symbolicvar,replacedvar)
				exec(code)
			return
    	
		#import hashlib
		#hashlib.sha256 = self._hookComplexFuncs(hashlib.sha256, self.gethookPaths, 'hexdigest', 'in2')
		#return

	def processhookFuncs(self, function, location, symbolicvar, replacedvar):
		@functools.wraps(function)
		def run(*args, **kwargs):
			returnVal = function(*args, **kwargs)
			if location[0] == 'args':
				target = tuple(args)[location[1]]
			elif location[0] == 'kwargs':
				target = kwargs[location[1]]
			elif location[0] == 'returnVal':
				if isinstance(returnVal, tuple):
					target = copy.deepcopy(returnVal[location[1]])
				else:
					target = copy.deepcopy(returnVal)

			for index in range(len(replacedvar)):
				try:
					if isinstance(target, dict):
						target[replacedvar[index]] = dict(self.symbolic_inputs)[symbolicvar[index]] 
					elif replacedvar[index] in dir(target):
						setattr(target, replacedvar[index], dict(self.symbolic_inputs)[symbolicvar[index]])
					else:
						print('Replaced parameter ' + replacedvar[index] + ' not found!')
				except:
					print('Replaced parameter ' + replacedvar[index] + ' not found!')
			
			if location[0] == 'returnVal':
				return target
			else:
				return returnVal
		return run

	def hookComplexFuncs(self, configPath):
		with open(configPath) as f:
			configJson = json.load(f)
			try:
				if configJson['complexfunc']['module'] == '' or configJson['complexfunc']['api'] == '' or configJson['complexfunc']['symbolicvar'] == '':
					return
			except:
				return
			code  = "import %s\n" % configJson['complexfunc']['module']
			code += "%s = self._hookComplexFuncs(%s, self.gethookPaths, '%s', '%s', '%s')" % (configJson['complexfunc']['api'], configJson['complexfunc']['api'], configJson['complexfunc']['api'], configJson['complexfunc']['property'], configJson['complexfunc']['symbolicvar'])
			exec(code)
			return

	def _hookComplexFuncs(self, function, prefixFunc, funcName, prop, inputVar):
		@functools.wraps(function)
		def run(*args, **kwargs):
			prefixFunc(function(*args, **kwargs), inputVar, prop, funcName)
			return function(*args, **kwargs)
		return run

	def gethookPaths(self, funcObj, inputVar, prop, func):
		if self.exploreDone:
			return
		inputVals = dict(self.symbolic_inputs)
		if inputVar not in inputVals:
			return

		if prop != None:
			prop_to_call = getattr(funcObj, prop)
			returnVal = prop_to_call()
		else:
			returnVal = funcObj
		
		conrVal = inputVals[inputVar].getConcrValue()

		noConfig = False
		if not (self.configPath and os.path.isfile(self.configPath)):
			noConfig = True
		else:
			with open(self.configPath) as f:
				configJson = json.load(f)
				if 'complexfunc' in configJson and 'symbolicvarVal' in configJson['complexfunc'] and 'symbolicvar' in configJson['complexfunc']:
					if conrVal not in configJson['complexfunc']['symbolicvarVal']:
						return
				else:
					noConfig = True
		if noConfig:
			return

		
		st = getSymbolic(conrVal)
		tmpS = st(inputVar, conrVal, ['==', st(inputVar, conrVal), conrVal])
		
		p = Predicate(tmpS, True)
		c = self.path.current_constraint.findChild(p)
		if c == None:
			c = self.path.current_constraint.addChild(p)
		
		self.path.current_constraint = c

		st = getSymbolic('stringval')
		tmpS = st(inputVar+'_'+re.sub(r'[^A-Za-z0-9\-_]', '_', func), returnVal, ['==', st(inputVar+'_'+re.sub(r'[^A-Za-z0-9\-_]', '_', func), returnVal), returnVal])
		c = self.path.current_constraint.findChild(Predicate(tmpS, True))
		if c == None:
			c = self.path.current_constraint.addChild(Predicate(tmpS, True))
		self.path.current_constraint = c
		return

	def printInputsConstraintsToSolve(self):
		for constraint in self.constraints_to_solve:
			if constraint.processed == True:
				continue
			print("constraint is: ", constraint)
			print("the input of the constraint is: ", constraint.inputs)

	def addConstraint(self, constraint):
		#if it's dict, delete the path with leaf node as ==, key#key, key
		if constraint.getLength() > 1:
			asserts, query = constraint.getDirectAssertAndQuery()
			if len(query.getVars()) > 0:
				queryVarName = query.getVars()[0]
				asserts = asserts.getVarStr()
				if asserts == '(== '+queryVarName+'#'+queryVarName+', '+queryVarName+')' and \
					query.getVarStr() != '(== '+queryVarName+'#'+queryVarName+', '+queryVarName+')':
					delPath = None
					for constr in self.constraints_to_solve:
						tmpasserts, tmpquery = constr.getDirectAssertAndQuery()
						if tmpquery.getVarStr() == '(== '+queryVarName+'#'+queryVarName+', '+queryVarName+')':
							delPath = constr
							break
					if delPath:
						self.constraints_to_solve.remove(delPath)
					constraint.removeDirectPartent()

		self.constraints_to_solve.append(constraint)
		# make sure to remember the input that led to this constraint
		constraint.inputs = self._getInputs()
		return

	def explore(self, max_iterations=0):
		self.exploreDone = False
		
		if self.configPath and os.path.isfile(self.configPath):
			with open(self.configPath) as f:
				configJson = json.load(f)
				if 'complexfunc' in configJson and 'symbolicvarVal' in configJson['complexfunc'] and 'symbolicvar' in configJson['complexfunc']:
					oldVal = self.symbolic_inputs.copy()
					for symVal in configJson['complexfunc']['symbolicvarVal']:
						updateFlag = self._updateSymbolicParameter(configJson['complexfunc']['symbolicvar'], symVal)
						self._oneExecution()
		
					self.symbolic_inputs = oldVal
		self._oneExecution()

		iterations = 1
		#When user press c after choosing option 2
		if max_iterations == -1:
			if not (self.configPath and os.path.isfile(self.configPath)):
				self.configPath = os.path.join(os.path.dirname(self._fileName), 'config.json')
			if os.path.isfile(self.configPath):
				with open(self.configPath) as f:
					configJson = json.load(f)
					if 'modify_hist' in configJson and configJson['modify_hist'][-1]['opr'] == 'bp':
						with open(configJson['modify_hist'][-1]['path']) as f:
							codeCont = f.readlines()
						del codeCont[configJson['modify_hist'][-1]['line']-1]
						with open(configJson['modify_hist'][-1]['path'], 'w') as f:
							f.write(''.join(codeCont))
						del configJson['modify_hist'][-1]
						json.dump(configJson, open(self.configPath, 'w'))
			sys.exit(1)
		elif max_iterations != 0 and iterations >= max_iterations:
			log.debug("Maximum number of iterations reached, terminating")
			self.exploreDone = True
			return self.generated_inputs, self.execution_return_values, self.path

		while not self._isExplorationComplete():
			selected = self.constraints_to_solve.popleft()
			if selected.processed:
				continue
			
			self._setInputs(selected.inputs)

			log.info("Selected constraint %s" % selected)
			#Get parents and last predicts
			asserts, query = selected.getAssertsAndQuery()

			model = None
			model = self.check_dict(asserts,query)
			cvc_expr = ''
			
			try:
				if model == None and str(type(self.solver)).find('CVCWrapper') != -1:
					model, cvc_expr = self.solver.findCounterexample(asserts, query)
				elif model == None:
					model = self.solver.findCounterexample(asserts, query)
					cvc_expr = None
			except:
				print("unexpected error occured when finding counterexample")
				log.critical("unexpected error occured when finding counterexample")
				model = None
			
			if model == None:
				continue
			else:
				updateFlag = False
				for name in model.keys():
					updateFlag = self._updateSymbolicParameter(name,model[name])
					if not updateFlag:
						continue
				if not updateFlag:
					continue
			self._oneExecution(selected, cvc_expr)

			iterations += 1
			self.num_processed_constraints += 1

			if max_iterations != 0 and iterations >= max_iterations:
				log.info("Maximum number of iterations reached, terminating")
				break

		self.exploreDone = True
		return self.generated_inputs, self.execution_return_values, self.path

	def print_dict(self, v, name, prefix=''):
		if isinstance(v, dict):
			for k, v2 in v.items():
				p2 = "{}['{}']".format(prefix, k)
				if name == 'None':
					target = open('prefix.txt', 'w')
					target.write(p2)
					target.close()

					return 
				self.print_dict(v2, name, p2)
		elif isinstance(v, list):
			for i, v2 in enumerate(v):
				p2 = "{}[{}]".format(prefix, i)
		else:
			if str(name) in prefix:
				target = open('prefix.txt', 'w')
				target.write(prefix)
				target.close()
				return 
	
	def update_nested_dict(self, in_dict, key_list, value, key_to_insert= None):
		if len(key_list) == 1:
			if type(value) == str and value == 'dict_action pop':
				in_dict.pop(key_list[0])
			elif type(value) == str and value.find('dict_action push')==0:
				if key_to_insert == 'expires_in':
					in_dict[key_to_insert] = 3600
				if key_to_insert == 'expires_at':
					in_dict[key_to_insert] = 3600
				else:
					in_dict[key_to_insert] = '3600'
			else:
				in_dict[key_list[0]] = value
			return 
		else:
			self.update_nested_dict(in_dict[key_list[0]], key_list[1:],value)

	def get_dict(self, expr):
		if isinstance(expr, dict):
			return expr
		elif isinstance(expr, list):
			for subexpr in expr:
				if not isinstance(subexpr, list) and not isinstance(subexpr, dict):
					continue
				return self.get_dict(subexpr)

	def check_dict(self, asserts, query):
		model = dict()
		#Boolean variable is None
		if query.symtype.expr == None:
			return None
		if len(query.symtype.expr) != 3:
			print('do not consider the case where len(query.symbolic_type.expr)!=3 '+ query.symtype.toString())
			raise ValueError
		else:
			op = query.symtype.expr[0]
			if op.find('dict') == 0:
				if op == 'dict_in':
					#negate query
					if query.result == True:
						model[query.symtype.expr[2]] = 'dict_action pop'
					else:
						dict_to_update = self.get_dict(query.symtype.expr[1])
						try:
							first_key = list(dict_to_update.keys())[0]
						except IndexError:
							first_key = None
						model[query.symtype.expr[2]] ='dict_action push siblingkey '+ str(first_key)
					return model
			else:
				return None

	# ---- private ---- #
	def _iter_dicts(self, iter_dict, name, val, keypath=""):
		# Another way is to use self.symbolic_inputs[name].expr to understand the operations on the dict
		for key, value in iter_dict.items():
			if isinstance(value, dict):
				return self._iter_dicts(value, name, val, keypath+"_."+str(key))
			elif key == name:
					st = getSymbolic(val)
					val = st(name,val)
					iter_dict[key] = val
					keypath = keypath +"_."+str(key)
					return keypath
					break
			else:
				continue

	def _updateSymbolicParameter(self, name, val):
		try:
			self.symbolic_inputs[name] = self.invocation.createArgumentValue(name,val)
			return True
		except KeyError:
			prefix = ''
			if type(val) == str and val.find('dict_action push') == 0:
				# Name is not in the dict. We use its siblingkey instead.
				siblingkey = val.split(' ')[-1]
				if siblingkey == 'None':
					self.print_dict(self.symbolic_inputs, 'None' , prefix)
				else:
					self.print_dict(self.symbolic_inputs, siblingkey, prefix)
			else:
				self.print_dict(self.symbolic_inputs, name, prefix)

			try:
				with open ("prefix.txt", "r") as myfile:
					prefix=myfile.readline()
				import os
				os.remove("prefix.txt")
			except FileNotFoundError:
				print("cannot find %s and %s", str(name), str(val))
				return False

			prefix = prefix.split(']')
			ordered_key_list = list()
			for item in prefix:
				if len(item) > 0:
					ordered_key_list.append(item.strip('[').strip('\''))

			import copy
			tmp =self.symbolic_inputs[ordered_key_list[0]].getConcrValue2()
			tmp_symbolic_inputs = copy.deepcopy(tmp)
			try:
				self.update_nested_dict(tmp_symbolic_inputs, ordered_key_list[1:], val, name)
			except IndexError:
				# empty dict
				self.update_nested_dict(tmp_symbolic_inputs, ordered_key_list[0:], val, name)

			name = ordered_key_list[0]
			val = tmp_symbolic_inputs

			if isinstance(self.symbolic_inputs[name], dict):
				self.symbolic_inputs[name].SymbolicValue()

			self.symbolic_inputs[name] = self.invocation.createArgumentValue(name,val)
			return True

	def _getInputs(self):
		return self.symbolic_inputs.copy()

	def _setInputs(self,d):
		self.symbolic_inputs = d

	def _isExplorationComplete(self):
		num_constr = len(self.constraints_to_solve)
		if num_constr == 0:
			log.info("Exploration complete")
			return True
		else:
			log.info("%d constraints yet to solve (total: %d, already solved: %d)" % (num_constr, self.num_processed_constraints + num_constr, self.num_processed_constraints))
			return False

	def _getConcrValue(self,v):
		if isinstance(v,SymbolicType):
			return v.getConcrValue()
		else:
			return v

	def _recordInputs(self):
		args = self.symbolic_inputs
		inputs = [ (k,self._getConcrValue(args[k])) for k in args ]
		self.generated_inputs.append(inputs)
		if self.numofExecution == 1:
			print("Initial inputs to the program are:")
			print(inputs)
		else:
			print("Inputs to the program are:")
			print(inputs)
		
	def _oneExecution(self, expected_path=None, cvc_expr = ""):
		self.numofExecution += 1
		self._recordInputs()
		
		self.path.reset(expected_path)

		try:
			ret = self.invocation.callFunction(self.symbolic_inputs)
		except BdbQuit as e:
			#When user press q after choosing option 2
			if not (self.configPath and os.path.isfile(self.configPath)):
				self.configPath = os.path.join(os.path.dirname(self._fileName), 'config.json')
			if os.path.isfile(self.configPath):
				with open(self.configPath) as f:
					configJson = json.load(f)
					if 'modify_hist' in configJson and configJson['modify_hist'][-1]['opr'] == 'bp':
						with open(configJson['modify_hist'][-1]['path']) as f:
							codeCont = f.readlines()
						del codeCont[configJson['modify_hist'][-1]['line']-1]
						with open(configJson['modify_hist'][-1]['path'], 'w') as f:
							f.write(''.join(codeCont))
						del configJson['modify_hist'][-1]
						json.dump(configJson, open(self.configPath, 'w'))
			sys.exit(1)
		except Exception as e:
			trackTrace = []
			for frame in traceback.extract_tb(sys.exc_info()[2]):
				codefname,codelineno,codefn,codemtext = frame
				trackTrace.append(codefname+'['+str(codelineno)+']: Function '+codefn+' '+codemtext)
			print(colored('Exception raised at the following localtion \n%s' % trackTrace[-1], 'red'))
			
			userChoice = self._askForChoice(trackTrace)

			if userChoice == '1':
				if not (self.configPath and os.path.isfile(self.configPath)):
					self.configPath = os.path.join(os.path.dirname(self._fileName), 'config.json')
				if not os.path.isfile(self.configPath):
					configJson = {}
				else:
					with open(self.configPath) as f:
						configJson = json.load(f)

				if 'modify_hist' not in configJson:
					configJson['modify_hist'] = []
				#Start changing the code
				with open(codefname) as f:
					codeCont = f.readlines()
				
				targetLine = re.split(r'([^\s])', codeCont[codelineno-1], maxsplit=1)
				codeCont[codelineno-1] = targetLine[0]+'#'+targetLine[1]+''.join(targetLine[2:])
				codeCont.insert(codelineno, targetLine[0]+'pass\n')
				configJson['modify_hist'].append({'path':codefname, 'line':codelineno, 'opr':'modify', 'orig':''.join(targetLine), 'new':codeCont[codelineno-1]})
				configJson['modify_hist'].append({'path':codefname, 'line':codelineno+1, 'opr':'add', 'orig':'', 'new':codeCont[codelineno]})
				print(colored('The following modification has been deployed:\nIn File %s \nChange line[%d] %s to %s\nIn File %s \nAdd line[%d] %s' % (codefname, codelineno, ''.join(targetLine), codeCont[codelineno-1], codefname, codelineno+1, codeCont[codelineno])))
				json.dump(configJson, open(self.configPath, 'w'))
				with open(codefname, 'w') as f:
					f.write(''.join(codeCont))

				print('Restart testing ... ')
				os.execl(sys.executable, sys.executable, *sys.argv)
			elif userChoice == '2':
				if not (self.configPath and os.path.isfile(self.configPath)):
					self.configPath = os.path.join(os.path.dirname(self._fileName), 'config.json')
				if not os.path.isfile(self.configPath):
					configJson = {}
				else:
					with open(self.configPath) as f:
						configJson = json.load(f)

				if 'modify_hist' not in configJson:
					configJson['modify_hist'] = []

				with open(codefname) as f:
					codeCont = f.readlines()
				loopline = codelineno
				childIndent = re.split(r'([^\s])', codeCont[loopline-1], maxsplit=1)[0]
				#get parent indent and insert a breakpoint
				loopline -= 1
				parentLine = None
				while loopline > 0:
					parentIndent = re.split(r'([^\s])', codeCont[loopline-1], maxsplit=1)[0]
					if len(parentIndent) < len(childIndent):
						parentLine = loopline
						break
					loopline -= 1
				if parentLine == None:
					print(colored('Can not locate its parent block', 'red'))
					return
				#Insert a breakpoint above its parent
				codeCont.insert(parentLine-1, parentIndent+'import pdb; pdb.set_trace()\n')
				with open(codefname, 'w') as f:
					f.write(''.join(codeCont))

				configJson['modify_hist'].append({'path':codefname, 'line':parentLine, 'opr':'bp', 'orig':'', 'new':parentIndent+'import pdb; pdb.set_trace()\n'})
				json.dump(configJson, open(self.configPath, 'w'))
				if '-m' not in sys.argv:
					sys.argv.append('-m')
					sys.argv.append('-1')
				else:
					sys.argv[sys.argv.index('-m')+1] = '-1'
				os.execl(sys.executable, sys.executable, *sys.argv)
			elif userChoice == '4':
				ret = None		
			elif userChoice == '5':
				sys.exit(0)
				return

		print('Function return %s \n************************************************************' % ret)
		self.execution_return_values.append(ret)
		if self.path.current_constraint.returnResult == None:
			self.path.current_constraint.returnResult = ret


	def _askForChoice(self, callTrace):
		while True:
			userChoice = input('You have the following options: \n1: Suppress this exception for later execution and rerun the program; \n2: Run with normal input and show me the correct value; \n3: Show me the function call stack;\n4: Return None and continue the execution;\n5: Quit now and I will manually change the code;\nPlease choose number 1, 2, 3, 4 or 5: ')
			userChoice = userChoice.strip()
			if userChoice == '3':
				print(colored('\n'.join(callTrace), 'green'))
			elif userChoice in ['1', '2', '4', '5']:
				return userChoice
