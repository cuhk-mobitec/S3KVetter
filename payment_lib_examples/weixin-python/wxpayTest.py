# -*- coding: utf-8 -*-
import pdb, time, wx_utils
from symbolic.args import *
from weixin.pay import WeixinPay, WeixinPayError

app_id = 'wxd678efh567hg6787'
mch_id = '1230000109'
api_key = '192006250b4c09247ec02edce69f6a2d'
total_fee = 10
notify_url = 'http://www.weixin.qq.com/wxpay/pap.php'
out_trade_no = '1409811653'

def generate_order(total_fee1):
	global total_fee

	if total_fee != total_fee1:
		total_fee = total_fee1

def verify_notification(data):
	global app_id, mch_id, total_fee, out_trade_no

	if not ((data['appid'] == app_id) and (data['mch_id'] == mch_id) and (data['total_fee'] == total_fee) and (data['out_trade_no'] == out_trade_no)):
		return False
	else:
		return True

@symbolic(sequential_notfied = 1, active_query = 1, total_fee1 = 10, total_fee2 = 10, app_id2 = 'wxd678efh567hg6787', mch_id2 = '1230000109', out_trade_no2 = '1409811653', transaction_id2 = '1004400740201409030005092168')
def wxpayTest(sequential_notfied, active_query, total_fee1, total_fee2, app_id2, mch_id2, out_trade_no2, transaction_id2):
	global app_id, mch_id, api_key, total_fee, notify_url, out_trade_no

	wxServer = wx_utils.wx_server(app_id, mch_id, api_key)

	generate_order(total_fee1)
	wx_pay = WeixinPay(app_id, mch_id, api_key, notify_url)
	try:
		order = wx_pay.unified_order(trade_type='APP', body='iPad Mini in white with 16G memory', total_fee=total_fee, out_trade_no=out_trade_no)
	except:
		order = wxServer.app_unified_order(total_fee, out_trade_no)
	try:
		if not wx_pay.check(order):
			print('Fail to verify the signature from WeChat server!')
			return 0
	except:
		return 0

	paymentRequest = {'appid': app_id, 'partnerid': mch_id, 'prepayid': order['prepay_id'], 'package': 'Sign=WXPay', 'timestamp': str(int(time.time())), 'noncestr': wx_pay.nonce_str}
	sign = wx_pay.sign(paymentRequest)
	paymentRequest['sign'] = sign

	xml = wxServer.asynchronous_notification(paymentRequest)

	if sequential_notfied: # asynchronous notification arrives before the synchronous one
		data = wx_pay.to_dict(xml)
		if not wx_pay.check(data):
			return 0
		if verify_notification(data):
			return 1
		else:
			return 0
	elif active_query: # actively query the cashier server
		try:
			query_response = pay.order_query(out_trade_no=out_trade_no)
		except:
			query_response = wxServer.order_query(out_trade_no=out_trade_no)
			if not query_response:
				return 0
			else:
				if verify_notification(query_response):
					return 1
				else:
					return 0	
	else:
		return 0