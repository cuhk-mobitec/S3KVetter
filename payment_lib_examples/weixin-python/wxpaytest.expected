(=> (and (or (= sequential_notfied 1) (= active_query 1)) (not (= total_fee1 20))) (= returnVal 0))
(=> (and (or (= sequential_notfied 1) (= active_query 1)) (not (= total_fee2 20))) (= returnVal 0))
(=> (and (or (= sequential_notfied 1) (= active_query 1)) (not (= app_id2 "wxd678efh567hg6787"))) (= returnVal 0))
(=> (and (or (= sequential_notfied 1) (= active_query 1)) (not (= mch_id2 "1230000109"))) (= returnVal 0))
(=> (and (or (= sequential_notfied 1) (= active_query 1)) (not (= out_trade_no2 "1409811653"))) (= returnVal 0))